core/src/oracles.ts

```diff
@@ -0,0 +1,135 @@
+        oracle: OracleCurrentAndNextPrices,
+        provider: ethers.providers.JsonRpcProvider,
+        oracleAddress: string
+    ) => {
+        /**
+         * Determine if the current price can be fetched from the contract,
+         * If yes, fetch the price by direct memory access (via slot address)
+         * The price is stored in the same slot as a validity marker:
+         *   - split the received value into validity marker part and price part
+         * If the price is valid - return it, otherwise provide a NaN
+         **/
+        const currentPriceFeed = await provider.getStorageAt(oracleAddress, oracle.currentPriceSlotAddress);
+        const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+        const isPriceValid = parseInt(currentPriceFeed.substring(0, valueSplitPosition), 16) === 1;
+        return isPriceValid
+            ? new BigNumber(`0x${currentPriceFeed.substring(valueSplitPosition)}`)
+            : new BigNumber(NaN);
+    },
+};
+const getCurrentOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    return currentPriceExtractors[oracle.type](oracle, provider, oracleAddress);
+};
+
+const getNextOraclePriceChange = async (
+    collateralConfig: CollateralConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    let nextPriceChange: Date = new Date(NaN);
+    if (collateralConfig.oracle.hasDelay) {
+        const osmContractInterface = await getContractInterfaceByName('OSM');
+        const osmContract = new ethers.Contract(oracleAddress, osmContractInterface, provider);
+        const lastPriceUpdateAsHex = (await osmContract.zzz())._hex;
+        const priceUpdateFrequencyInSeconds = await osmContract.hop();
+        const lastPriceUpdateTimestampInSeconds = new BigNumber(lastPriceUpdateAsHex).toNumber();
+        nextPriceChange = new Date((lastPriceUpdateTimestampInSeconds + priceUpdateFrequencyInSeconds) * 1000);
+    }
+    return nextPriceChange;
+};
+
+const _getOsmPrices = async (
+    network: string,
+    oracleAddress: string,
+    collateralType: CollateralType
+): Promise<OraclePrices> => {
+    const provider = await getProvider(network);
+    const collateralConfig = COLLATERALS[collateralType];
+
+    const nextPrice = await getNextOraclePrice(collateralConfig.oracle, provider, oracleAddress);
+    const currentPrice = await getCurrentOraclePrice(collateralConfig.oracle, provider, oracleAddress);
+    const nextPriceChange = await getNextOraclePriceChange(collateralConfig, provider, oracleAddress);
+
+    const currentUnitCollateralPrice = new BigNumber(currentPrice).shiftedBy(-WAD_NUMBER_OF_DIGITS);
+    const nextUnitCollateralPrice = new BigNumber(nextPrice).shiftedBy(-WAD_NUMBER_OF_DIGITS);
+
+    return {
+        currentUnitPrice: currentUnitCollateralPrice,
+        nextUnitPrice: nextUnitCollateralPrice || new BigNumber(NaN),
```

Comment:
In which case this will be falsy? I can see that `nextUnitCollateralPrice` will always contain `new BigNumber()`

Reply:
removed the right part